polus-graph-pyramid-builder-plugin/src/main.py

```diff
@@ -62,66 +73,415 @@ def load_csv(fpath):
+                                                                       bin_stats['max'][neg2posnames])
+    yaxis = yaxis + yvalues
+    alphavals = alphavals + alphasneg2pos
+    column_bin_sizes = column_bin_sizes + commonratiosneg2pos
+    neg2posdf.reset_index(drop = True, inplace = True)
+
+    # NEW DATA FRAME DROPS COLUMNS THAT HAS A BIN WIDTH VALUE OF ZERO
+    data = pandas.concat([positivedf, negativedf, neg2posdf], axis=1)
+    column_names = data.columns
+
+    bin_stats = {'size': data.shape,
+                 'min': data.min(),
+                 'max': data.max()}
+
+    nfeats = bin_stats['size'][1] 
+    datalen = bin_stats['size'][0]
+
+
+    data_ind = pandas.notnull(data)  # Handle NaN values
+    data[~data_ind] = 255          # Handle NaN values
+    data = data.astype(np.uint16) # cast to save memory
+    data[data==bincount] = bincount - 1         # in case of numerical precision issues
+
+    nrows = data.shape[0]
+    if nrows < 2**8:
+        dtype = np.uint8
+    elif nrows < 2**16:
+        dtype = np.uint16
+    elif nrows < 2**32:
+        dtype = np.uint32
+    else:
+        dtype = np.uint64
+    
+    totalgraphs = (scipy.special.factorial(nfeats))/(2*(scipy.special.factorial(nfeats-2)))
+    totalgraphs = totalgraphs.astype(np.uint32)
+    bins = np.zeros((totalgraphs,bincount,bincount),dtype=dtype)
+    log_index = []
+    log_dict = {}
+
+    # Create a linear index for feature bins
+    i = 0
+    for feat1 in range(nfeats):
+        name1 = column_names[feat1]
+        feat1_tf = data[name1] * bincount
+
+        for feat2 in range(feat1 + 1, nfeats):
+            log_dict[(feat1, feat2)] = i
+            Datapoints.append([])
+            Ind2.append([])
+            Ind2zero.append([])
+            Position.append([])
+            name2 = column_names[feat2]
+            feat2_tf = data[name2]
+            feat1_tf = feat1_tf[data_ind[name1] & data_ind[name2]]
+            feat2_tf = feat2_tf[data_ind[name1] & data_ind[name2]]
+                      
+            if feat2_tf.size<=1:
+                continue
+            
+            # sort linear matrix indices
+            SortedFeats = np.sort(feat1_tf + feat2_tf)
+            # Do math to get the indices
+            ind2 = np.diff(SortedFeats)                       
+            ind2 = np.nonzero(ind2)[0]                       # nonzeros are cumulative sum of all bin values
+            ind2 = np.append(ind2,SortedFeats.size-1)
+            # print(feat2_sort.shape)
+            rows = (SortedFeats[ind2]/bincount).astype(np.uint8)   # calculate row from linear index
+            cols = np.mod(SortedFeats[ind2],bincount)              # calculate column from linear index
+            counts = np.diff(ind2)                           # calculate the number of values in each bin
+            bins[i,rows[0],cols[0]] = ind2[0] + 1
+            bins[i,rows[1:],cols[1:]] = counts
+            log_index.append([feat1,feat2])
```

Comment:
Rather than duplicating this code for the linear and log bin count methods, it might be better to just create a method that takes data where bins have been assigned generate the 2d bin data. That way if there is a memory issue (like we experience recently) then the code for this identical block of code only needs to be changed in one spot.